pkg/controllers/kepler.go

```diff
@@ -0,0 +1,300 @@
+		logger.V(6).Info("Get Error ", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	if kepler == nil {
+		// no kepler found , so stop here
+		logger.V(6).Info("Kepler Nil")
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: have admission webhook decline all but `kepler` instance
+	if kepler.Name != "kepler" {
+		return r.setInvalidStatus(ctx, req)
+	}
+
+	logger.V(6).Info("Running sub reconcilers", "kepler", kepler.Spec)
+
+	result, recErr := r.runKeplerReconcilers(ctx, kepler)
+	updateErr := r.updateStatus(ctx, req, err)
+
+	if recErr != nil {
+		return result, recErr
+	}
+	return result, updateErr
+}
+
+func (r KeplerReconciler) runKeplerReconcilers(ctx context.Context, kepler *v1alpha1.Kepler) (ctrl.Result, error) {
+
+	reconcilers := r.reconcilersForKepler(kepler)
+	r.logger.V(6).Info("renconcilers ...", "count", len(reconcilers))
+
+	return reconciler.Runner{
+		Reconcilers: reconcilers,
+		Client:      r.Client,
+		Scheme:      r.Scheme,
+		Logger:      r.logger,
+	}.Run(ctx)
+}
+
+func (r KeplerReconciler) getKepler(ctx context.Context, req ctrl.Request) (*v1alpha1.Kepler, error) {
+	logger := r.logger
+
+	kepler := v1alpha1.Kepler{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &kepler); err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(3).Info("kepler could not be found; may be marked for deletion")
+			return nil, nil
+		}
+		logger.Error(err, "failed to get kepler")
+		return nil, err
+	}
+
+	return &kepler, nil
+}
+
+func (r KeplerReconciler) updateStatus(ctx context.Context, req ctrl.Request, recErr error) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+
+		kepler, _ := r.getKepler(ctx, req)
+		// may be deleted
+		if kepler == nil || !kepler.GetDeletionTimestamp().IsZero() {
+			// retry since some error has occurred
+			r.logger.V(6).Info("Reconcile has deleted kepler; skipping update")
+			return nil
+		}
+
+		if recErr != nil {
+			kepler.Status.Conditions = []metav1.Condition{{
+				Type:               v1alpha1.ConditionReconciled,
+				Status:             metav1.ConditionTrue,
+				ObservedGeneration: kepler.Generation,
+				LastTransitionTime: metav1.Now(),
+				Reason:             v1alpha1.ReconciledReasonError,
+				Message:            recErr.Error(),
+			}}
+		} else {
+			kepler.Status.Conditions = []metav1.Condition{{
+				Type:               v1alpha1.ConditionReconciled,
+				Status:             metav1.ConditionTrue,
+				ObservedGeneration: kepler.Generation,
+				Reason:             v1alpha1.ReconciledReasonComplete,
+				Message:            "Reconcile succeeded",
+				LastTransitionTime: metav1.Now(),
+			}}
+		}
+		return r.Client.Status().Update(ctx, kepler)
+
+	})
+}
+
+func (r KeplerReconciler) reconcilersForKepler(k *v1alpha1.Kepler) []reconciler.Reconciler {
+	// start with namespace
+	rs := []reconciler.Reconciler{}
+
+	if cleanup := !k.DeletionTimestamp.IsZero(); cleanup {
+		rs = append(rs, deletersForResources(components.NewKeplerNamespace())...)
```

Comment:
Won't this delete Operator as well from Namespace?

Reply:
No, it wont .. Operator gets installed to `openshift-operators` and kepler-operator creates another namespace for deploying its resources which what is being deleted here.